lib/autocheck/include/autocheck/apply.hpp

```diff
@@ -23,14 +23,6 @@ namespace autocheck {
     typedef typename std::result_of<F(Args&...)>::type type;
   };
 
-  template <typename F, typename Tuple>
```

Comment:
why is this not needed? I don't see this change in upstream... and I actually don't understand how it works without this.

Also, we should avoid deviating from upstream if we can.

Reply:
In C++17 there's a `std::apply` which looks to me as though it does the same thing (in different words):

https://en.cppreference.com/w/cpp/utility/apply

So with the `autocheck` `apply` deleted, we just use the STL one, whereas with it present, we get:

```
In file included from crypto/test/CryptoTests.cpp:17:
In file included from ../lib/autocheck/include/autocheck/autocheck.hpp:4:
../lib/autocheck/include/autocheck/check.hpp:48:12: error: call to 'apply' is ambiguous
      if (!apply(prop, args.ref())) {
           ^~~~~
../lib/autocheck/include/autocheck/check.hpp:59:5: note: in instantiation of function template specialization 'autocheck::check<std::vector<unsigned char>, (lambda at crypto/test/CryptoTests.cpp:62:9), autocheck::arbitrary<autocheck::generator<std::vector<unsigned char>>>>' requested here
    check<Args...>(std::move(prop),
    ^
crypto/test/CryptoTests.cpp:61:16: note: in instantiation of function template specialization 'autocheck::check<std::vector<unsigned char>, (lambda at crypto/test/CryptoTests.cpp:62:9)>' requested here
    autocheck::check<std::vector<uint8_t>>(
               ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/tuple:1729:5: note: candidate function [with _Fn = (lambda at crypto/test/CryptoTests.cpp:62:9) &, _Tuple = std::tuple<std::vector<unsigned char>> &]
    apply(_Fn&& __f, _Tuple&& __t)
    ^
../lib/autocheck/include/autocheck/apply.hpp:28:3: note: candidate function [with F = (lambda at crypto/test/CryptoTests.cpp:62:9), Tuple = std::tuple<std::vector<unsigned char>> &]
  apply(F f, Tuple&& args) {
  ^
```

Other options besides just deleting that include updating to upstream en masse and isolating just whatever they did to fix this and pulling that change.

Reply:
Yeah we should first try upstream, and if it doesn't work look into rolling our own.

It looks to me that this entire file should go away (as everything in here looks to exist as building blocks for `apply`), potentially more.

Reply:
Upstream has not altered this file since 2015. Unfortunately it's fairly dormant. I don't think it compiles on C++17 by default.

Reply:
OK, so I guess we can just make it work for C++ 17. From what I see it's basically deleting this file and use `std::apply` instead

Reply:
@MonsieurNicolas done